mesh/mesh.go

```diff
@@ -406,19 +399,43 @@ func (msh *Mesh) pushLayersToState(ctx context.Context, oldVerified, newVerified
 	// we never reapply the state of oldVerified. note that state reversions must be handled separately.
 	for layerID := oldVerified.Add(1); !layerID.After(newVerified); layerID = layerID.Add(1) {
 		l, err := msh.GetLayer(layerID)
-		// TODO: propagate/handle error
-		if err != nil || l == nil {
+		if err != nil {
 			logger.With().Error("failed to get layer", layerID, log.Err(err))
-			return
+			return err
 		}
+
 		validBlocks, invalidBlocks := msh.BlocksByValidity(l.Blocks())
-		msh.updateStateWithLayer(ctx, types.NewExistingLayer(layerID, validBlocks))
+		var (
+			applied    *types.Block
+			notApplied = invalidBlocks
+			blocks     = types.SortBlocks(validBlocks)
+		)
+
+		if len(blocks) > 0 {
+			// when tortoise verify multiple blocks in the same layer, we only apply one with the lowest
+			// lexicographical sort order
+			applied = blocks[0]
+			if len(blocks) > 1 {
+				notApplied = append(notApplied, blocks[1:]...)
+			}
+		}
+
+		if err = msh.updateStateWithLayer(ctx, layerID, applied); err != nil {
```

Comment:
is it expected that applied will be nil is layer is empty?

Reply:
yes. if the hare output an empty set, then it's likely that the tortoise will not verify any block. i.e. no blocks in that layer has contextual validity bit set to true.